### PR TITLE
PERF: Faster comparisons of indexes when compared to self

### PR DESCRIPTION
**Pull request by topper-123**
Examples:

```python
>>> rng = pd.RangeIndex(100_000)
>>> %timeit rng == rng
429 µs ± 96.9 µs per loop  # master
14.1 µs ± 93.8 ns per loop  # this PR
>>> idx = rng.astype(int)
>>> %timeit idx == idx
347 µs ± 1.42 µs per loop  #master
14.1 µs ± 91.4 ns per loop  # this PR
>>> idx = rng.astype(object)
>>> %timeit idx == idx
5.97 ms ± 12.7 µs per loop  # master
3.72 ms ± 28.4 µs per loop  # this PR
```

### CODE REVIEW

**Comment by jbrockmendel** [comment]
that works for int dtype, but what about for dtypes that have NAs?

**Comment by topper-123** [comment]
Hmm, that’s a good point. I’ll see if anything can be done about that.

**Comment by twoertwein** [comment]
you could return `~isna()` for equal and `isna()` for not equal (if that is still faster)?

**Comment by jbrockmendel** [comment]
> you could return ~isna() for equal and isna() for not equal (if that is still faster)?

even then you'd need to watch out for pd.NA

I think just check for _can_hold_na.

Also for RangeIndex could compare the range objects instead of (i think) casting to ndarray

**Comment by topper-123** [comment]
I've made an update. I can't find the tests for index comparisons, they clearly need some tests for comparisons with indexes that contain nans. My first proposal obviously shouldn't have passed.

Anyonw know where they're located?

**Comment by jbrockmendel** [comment]
> Anyonw know where they're located?

tests.arithmetic

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -5363,6 +5363,24 @@ def _cmp_method(self, other, op):
         """
         Wrapper used to dispatch comparison operations.
         """
+        from .multi import MultiIndex
+
+        if self.is_(other) and not isinstance(self, MultiIndex):
+            # short-circuit when other is same as self
+            # if we're comparing equality, return an np.ones array, else an np.zeros arr
+            bool_arr_type, na_bool = {
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:5371]: the pattern we use elsewhere is more like

```
res = np.empty(self.shape, dtype=bool)
res[:] = (op.__name__ in ["eq", "le", "ge"]
if self._can_hold_na:
   ...
```

**Comment by jbrockmendel** [comment]
would this let us remove NumericIndex._cmp_method?

**Comment by jbrockmendel** [comment]
@topper-123 can you merge master

**Comment by jreback** [comment]
@topper-123 still worth it here? can you merge master

**Comment by topper-123** [comment]
This has already been fixed, e.g. the last example:

```python
>>> idx = rng.astype(object)
>>> %timeit idx == idx
5.97 ms ± 12.7 µs per loop  # when this PR was opened
53.9 µs ± 9.8 µs per loop  # master
```

So I'll just close this one.

**Pull request closed without merging**